simulator/func_sim/func_sim.cpp

```diff
@@ -75,6 +75,55 @@ void FuncSim<ISA>::run( const std::string& tr, uint64 instrs_to_run)
     }
 }
 
+template <typename ISA>
+void FuncSim<ISA>::gdb_load( const std::string &tr) {
+    mem->load_elf_file( tr);
+    sout << "MIPT-MIPS: Binary file " << tr << " loaded" << std::endl;
+    //TODO: print more useful info
+}
+
+template <typename ISA>
+void FuncSim<ISA>::gdb_prepare() {
+    PC = mem->startPC();
+    nops_in_a_row = 0;
+    sout << "MIPT-MIPS: prepared to run" << std::endl;
+    //TODO: print more useful info
+}
+
+template <typename ISA>
+void FuncSim<ISA>::gdb_resume( int steps) {
+    sout << "MIPT-MIPS: resuming, steps: " << steps << std::endl;
+    uint32 instrs_to_run = ((steps == 0) ? MAX_VAL32 : steps);
+    for ( uint32 i = 0; i < instrs_to_run; ++i) {
+        const auto& instr = step();
+        sout << instr << std::endl;
+        if ( instr.is_halt())
+            break;
+    }
+}
+
+template <typename ISA>
+size_t FuncSim<ISA>::gdb_mem_read (Addr addr, unsigned char *buf, size_t length) try {
+    size_t bytes_read = 0;
+    for (; bytes_read < length; bytes_read++)
```

Comment:
This should call FuncMemory functions `memcpy_host_to_guest` and `memcpy_guest_to_host`.
If one day we decide to optimize them with vectorization, we will do it in the one place.

We might need an exception-free wrappers though, they should be named `try_mem_read`.